File: Example/common/src/main/java/com/example/common/okhttp/download/DownloadRunnable.java
package com.example.common.okhttp.download;


import android.os.Environment;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.util.Objects;

import okhttp3.Response;

/**
 * @author tengfei
 * date 2019/2/16 9:33 PM
 * email tengfeigo@outlook.com
 * description
 */
public class DownloadRunnable implements Runnable {
    private String mUrl;
    private int threadId;
    private long startSize;
    private long endSize;
    private final DownloadCallBack downloadCallBack;

    private static final int STATUS_DOWNLOADING = 0x001;
    private static final int STATUS_STOP = 0x002;
    private int mStatus = STATUS_DOWNLOADING;

    private long mProgress;

    public DownloadRunnable(String mUrl, int threadId, long startSize, long endSize, DownloadCallBack downloadCallBack) {
        this.mUrl = mUrl;
        this.threadId = threadId;
        this.startSize = startSize;
        this.endSize = endSize;
        this.downloadCallBack = downloadCallBack;
    }

    @Override
    public void run() {
        //只负责读写自己的内容
        InputStream inputStream = null;
        RandomAccessFile randomAccessFile = null;
        try {
            Response response = OkHttpManager.getInstance().syncResponse(mUrl, startSize, endSize);
            inputStream = Objects.requireNonNull(response.body()).byteStream();
            File file = FileManager.getInstance().getFile(mUrl);
            randomAccessFile = new RandomAccessFile(file, "rwd");
            randomAccessFile.seek(startSize);

            int len = 0;
            byte[] buffer = new byte[1024];
            while ((len = inputStream.read(buffer)) != -1) {
                if (mStatus == STATUS_STOP) {
                    break;
                }
                mProgress += len;
                randomAccessFile.write(buffer, 0, len);
            }
            downloadCallBack.onSuccess(file);
        } catch (IOException e) {
            downloadCallBack.onFailure(e);
        } finally {
            Utils.close(inputStream);
            Utils.close(randomAccessFile);
        }
    }

    public void stop() {
        mStatus = STATUS_STOP;
    }
}

File: OkHttpExample/app/src/main/java/com/example/okhttp/DownloadRunnable.kt
package com.example.okhttp

import okhttp3.Response
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.RandomAccessFile

/**
 * @author tengfei
 * date 2019/3/5 11:17 PM
 * email tengfeigo@outlook.com
 * description
 */
class DownloadRunnable(url: String, start: Long, end: Long, downloadCallBack: DownloadCallBack?) : Runnable {

    private var url: String? = null
    private var start: Long? = null
    private var end: Long? = null
    private var downloadCallBack: DownloadCallBack? = null

    init {
        this.url = url
        this.start = start
        this.end = end
        this.downloadCallBack = downloadCallBack
    }

    override fun run() {
        val file: File = FileManager.instance.createFile()
        var inputStream: InputStream? = null
        var randomAccess: RandomAccessFile? = null
        try {
            val response: Response = OkHttpManager.instance.syncRequest(url!!, start!!, end!!)
            inputStream = response.body()!!.byteStream()
            randomAccess = RandomAccessFile(file, "rwd")
            val buffer = ByteArray(1024 * 10)
            var len = 0
            while ((inputStream.read(buffer)) != -1) {
                len = inputStream.read(buffer)
                randomAccess.write(buffer, 0, len)
            }
            downloadCallBack!!.onSuccess(file)
        } catch (e: IOException) {
            downloadCallBack!!.onFailure(e)
        } finally {
            Utils.close(inputStream!!)
            Utils.close(randomAccess!!)
        }


    }
}